core/src/main/java/hudson/model/Queue.java

```diff
@@ -3009,4 +3010,48 @@ public static Queue getInstance() {
     public static void init(Jenkins h) {
         h.getQueue().load();
     }
+
+    /**
+     * Schedule <tt>Queue.save()</tt> call for near future once items change. Ignore all changes until the time the save takes place.
+     */
+    @Extension
+    @Restricted(NoExternalUse.class)
+    public static final class Saver extends QueueListener implements Runnable {
+        @VisibleForTesting
+        /*package*/ static /*final*/ int DELAY_SECONDS = 60;
+        private volatile Future<?> nextSave;
+
+        public Saver() {
+            // Save the queue on JVM shutdown
+            Runtime.getRuntime().addShutdownHook(new Thread(this, "Save Jenkins Queue on Shutdown"));
```

Comment:
Already saved via https://github.com/jenkinsci/jenkins/blob/08a2b1e4410cb5713f43227096780f0933b909d7/core/src/main/java/jenkins/model/Jenkins.java#L3500

Reply:
I did not find a code path proving JVM shutdown would always call `Jenkins#cleanUp()`. Can we rely it will happen?

Reply:
Based on my causal testing this is indeed unnecessary.